src/main/java/com/google/devtools/build/lib/bazel/repository/PatchUtil.java

```diff
@@ -0,0 +1,210 @@
+    List<String> originContent = readFile(oldFile);
+    Patch<String> patch = DiffUtils.parseUnifiedDiff(patchContent);
+    List<String> newContent = DiffUtils.patch(originContent, patch);
+    writeFile(newfile, newContent);
+  }
+
+  private static boolean isGitFormatPatch(List<String> patches) {
+    if (patches.size() > 4) {
+      return patches.get(0).startsWith("From")
+          && patches.get(1).startsWith("From:")
+          && patches.get(2).startsWith("Date:")
+          && patches.get(3).startsWith("Subject:");
+    }
+    return false;
+  }
+
+  /**
+   * Preprocess the original patch file.
+   * 1. Removing leading and trailing empty lines.
+   * 2. Removing ambiguous lines when dealing with git format patch.
+   * 3. Adding an extra line to make sure last chunk also gets applied later.
+   * @param patches The original patch file content
+   * @return The processed patch file content
+   */
+  private static List<String> preprocessPatch(List<String> patches) {
+    // Trimming empty lines from head and tail
+    int left = 0;
+    int right = patches.size() - 1;
+    while (left < patches.size() && patches.get(left).isEmpty()) {
+      left++;
+    }
+    while (right > 0 && patches.get(right).isEmpty()) {
+      right--;
+    }
+    if (left <= right) {
+      patches = patches.subList(left, right + 1);
+    } else {
+      patches = new ArrayList<>();
+    }
+
+    // If it's a git format patch, "--" is ambiguous. We don't know it belongs to a patch chunk or
+    // just the footer of the patch file.
+    // So removing the last 2 lines that looks like:
+    // --
+    // 2.18.0.rc1.244.gcf134e6275-goog
+    if (isGitFormatPatch(patches)) {
+      int size = patches.size();
+      if (patches.get(size - 2).trim().equals("--")) {
+        patches = patches.subList(0, size - 2);
+      }
+    }
+
+    // Adding an extra line to make sure last chunk also gets applied.
+    patches.add("$");
+    return patches;
+  }
+
+  /**
+   * Strip a number of leading components from a path
+   * @param path the original path
+   * @param strip the number of leading components to strip
+   * @return The stripped path
+   */
+  private static String stripPath(String path, int strip) {
+    int pos = 0;
+    while (pos < path.length() && strip > 0) {
+      if (path.charAt(pos) == '/') {
+        strip--;
+      }
+      pos++;
+    }
+    return path.substring(pos);
+  }
+
+  private static String getFilePath(String line, int strip, Path outputDirectory) {
+    // The line could look like:
+    // --- a/foo/bar.txt   2019-05-27 17:19:37.054593200 +0200
+    // +++ b/foo/bar.txt   2019-05-27 17:19:37.054593200 +0200
+    // If strip is 1, we want get the file path as <outputDirectory>/foo/bar.txt
+    String file = null;
+    line = line.split("\t")[0];
+    if (line.length() > 4) {
+      file = stripPath(line.substring(4), strip);
+      file = outputDirectory.getRelative(file).getPathString();
```

Comment:
That approach will break for patches generating or removing files, where the name description might look like

```
--- /dev/null
+++ b/src/main/java/com/google/devtools/build/lib/bazel/repository/PatchUtil.java
```

Reply:
Thanks for notice, will fix!